src/coprocessor/dag/batch/executors/util/aggr_executor.rs

```diff
@@ -59,20 +59,32 @@ pub trait AggregationExecutorImpl<Src: BatchExecutor>: Send {
     ) -> Result<()>;
 
     /// Returns the current number of groups.
+    ///
+    /// Note that this number can be inaccurate because it is a hint for the capacity of the vector.
     fn groups_len(&self) -> usize;
 
-    /// Iterates aggregate function states for each group.
+    /// Iterates aggregate function states for each available group.
     ///
     /// Implementors should call `iteratee` for each group with the aggregate function states of
     /// that group as the argument.
     ///
     /// Implementors may return the content of each group as extra columns in the return value
     /// if there are group by columns.
-    fn iterate_each_group_for_aggregation(
+    ///
+    /// Implementors should not iterate the same group multiple times for the same partial
+    /// input data.
+    fn iterate_available_groups(
         &mut self,
         entities: &mut Entities<Src>,
+        src_is_drained: bool,
         iteratee: impl FnMut(&mut Entities<Src>, &[Box<dyn AggrFunctionState>]) -> Result<()>,
     ) -> Result<Vec<LazyBatchColumn>>;
+
+    /// Returns whether we can now output partial aggregate results when the source is not drained.
+    ///
+    /// This method is called only when the source is not drained because aggregate result is always
+    /// ready if the source is drained and no error occurs.
+    fn is_partial_results_ready(&self) -> bool;
```

Comment:
Why not provide a default implementation for this method, and reduce repeated code in `Fast/Slow/Simple` aggregation implementations.

Reply:
See https://github.com/tikv/tikv/pull/4786#discussion_r287831683